sdk/e2e/02_update_router_invalid_test.py

```diff
@@ -0,0 +1,83 @@
+import os
+import logging
+
+import requests
+import turing
+
+from turing.router.config.experiment_config import ExperimentConfig
+from turing.router.config.resource_request import ResourceRequest
+from turing.router.config.enricher import Enricher
+from turing.router.config.common.env_var import EnvVar
+from turing.router.config.router_version import RouterStatus
+
+
+def test_update_router_invalid_config():
```

Comment:
My understanding was that pytest runs the tests in a random order. So I was expecting to see one of the solutions described here, to ensure order of execution: https://stackoverflow.com/questions/17571438/test-case-execution-order-in-pytest

How do we make sure that this test always runs after the first?

Reply:
I believe the tests within a module or a class run in a random order but module discovery seems to be something non-random; pytest appears to recursively look into each directory to collect individual test modules in an alphabetical order (doesn't say it explicitly[ here](https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html#conventions-for-python-test-discovery) but it kinda hints at it). 

In our case, I've named each file kinda alphabetically with `01_`, `02_`, ... `08_` to force an order during test collection and pytest doesn't seem to ever collect the tests in a different order other than the one specified 🤔

But yea arguably I think a cleaner and 'surer' way to ordering the tests would be to use the fixtures/hooks method suggested.

Reply:
I see.. Here's another SO answer that mentions that:
> The tests will most likely run in alphanumeric order based on the file name. 

https://stackoverflow.com/questions/44234414/which-file-is-executed-first-in-pytest/44234464#44234464

But it also highlights that parallelism could affect this behaviour. To be safe, since the tests are from different modules, shall we use fixtures to mark the dependency?

Reply:
Ok I've tried using fixtures to specify the test order as described in this response: https://stackoverflow.com/a/50575775. Unfortunately it doesn't work for chained fixtures, i.e. a test which itself is a fixture and uses another fixture (which is what we'd need since we need `08` to depend on `07`, which depends on `06`, ...), due to this bug specified in the warning of this section: https://stackoverflow.com/a/50575775:

> Warning
Note this mark has no effect in fixture functions. For example, this will not work as expected:
@pytest.mark.usefixtures("my_other_fixture")
@pytest.fixture
def my_fixture_that_sadly_wont_use_my_other_fixture():
    ...
Currently this will not generate any error or warning, but this is intended to be handled by [issue #3664](https://github.com/pytest-dev/pytest/issues/3664).

Instead, I've decided to use the `pytest_collection_modifyitems` hook recommended in `conftest.py`, which manually sorts out the collected test functions with respect to a fixed list of test names to ensure the correct test order during execution:

```python
TEST_ORDER = [
    "test_create_router",
    "test_update_router_invalid_config",
    "test_create_router_version",
    "test_undeploy_router",
    "test_deploy_router_invalid_config",
    "test_deploy_router_valid_config",
    "test_delete_router",
    "test_deploy_router_with_traffic_rules"
]

def pytest_collection_modifyitems(items):
    # modify test items in place to ensure that the e2e tests run in order
    item_mappings = {item.originalname: item for item in items}
    sorted_items = []
    for test_name in TEST_ORDER:
        sorted_items.append(item_mappings[test_name])
    items[:] = sorted_items
```

Reply:
I see. Cool, thanks for digging into this!